llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDKernelAttrs.cpp

```diff
@@ -12,14 +12,76 @@
 #include "llvm/SYCLLowerIR/ESIMD/LowerESIMD.h"
 #include "llvm/SYCLLowerIR/SYCLUtils.h"
 
-#include "llvm/IR/Module.h"
-#include "llvm/Pass.h"
-
 #define DEBUG_TYPE "LowerESIMDKernelAttrs"
 
 using namespace llvm;
 
 namespace llvm {
+
+// Checks if Call Instruction corresponds to InvokeSimd function call.
+bool isInvokeSimdBuiltinCall(const CallInst *CI) {
+  Function *F = CI->getCalledFunction();
+
+  return F && F->getName().startswith(esimd::INVOKE_SIMD_PREF);
+}
+
+// Receives as a parameter an address of memory location where address function
+// being traversed by traverseCallgraphUp is stored. Checks if this memory
+// location is used diectly or indirectly as an argument for invoke_simd.
+// Returns false if the function is used as an argument for
+// invoke_simd function call, true otherwise.
+bool checkFunctionAddressUse(const Value *address) {
```

Comment:
Looking into the code deeper, I think this part should be reworked. The overall approach can be:

assert(no stores to the address other than given function)
Loads = find all loads from `address`
for (Load : Loads) {
  Uses = collectUsesSkipThroughCasts();
  for (Use : Uses) {
    if (Use is parameter to invoke_simd) {
      return true;
    }
  }
  return false;
}

Current logic is very different. We also don't need recursion, so this can be inlined intp the caller.

Reply:
I think original `checkFunctionAddressUse` is borrowed from https://github.com/intel/llvm/blob/sycl/llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp#L102, whose purpose is quite different from what we need here. `collectPossibleStoredVals` is used to conservatively detect if optimization is possible, `checkFunctionAddressUse` must tell with 100% reliability that given address:
1) either contains only given function address or cannot contain it at all, anything in between should be an assert (as we lose reliability in this case)
2) if it contains only given function address, then the value stored (the function pointer) either can't appear as an argument of __builtin_invoke_simd.
In general case this is not possible, e.g. when user program uses some complex data flow on the function pointer before calling invoke_simd.

The more I think of this, the more I'm getting inclined to stop adding a call graph edge on any use of a function other than call.
In our case, if a user function can call ESIMD function via a pointer, we should require that this function is marked with SYCL_ESIMD_FUNCTION explicitly.

Reply:
As we talked, the idea is:
- When propagating sycl_explicit_simd attribute we never go up through non-call use of function pointers, which automatically covers invoke_simd
- Users must explicitly add sycl_explicit_simd attribute if function calls a sycl_explicit_simd function through pointer.